receiver/dcgmreceiver/scraper.go

```diff
@@ -90,44 +95,144 @@ func (s *dcgmScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+			s.mb.RecordGpuDcgmSmOccupancyDataPoint(now, metric.asFloat64())
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_PIPE_TENSOR_ACTIVE"]; ok {
+			s.mb.RecordGpuDcgmPipeUtilizationDataPoint(now, metric.asFloat64(), metadata.AttributeGpuPipeTensor)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_PIPE_FP64_ACTIVE"]; ok {
+			s.mb.RecordGpuDcgmPipeUtilizationDataPoint(now, metric.asFloat64(), metadata.AttributeGpuPipeFp64)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_PIPE_FP32_ACTIVE"]; ok {
+			s.mb.RecordGpuDcgmPipeUtilizationDataPoint(now, metric.asFloat64(), metadata.AttributeGpuPipeFp32)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_PIPE_FP16_ACTIVE"]; ok {
+			s.mb.RecordGpuDcgmPipeUtilizationDataPoint(now, metric.asFloat64(), metadata.AttributeGpuPipeFp16)
+		}
+		if metric, ok := metrics["DCGM_FI_DEV_ENC_UTIL"]; ok {
+			encUtil := float64(metric.asInt64()) / 100.0 /* normalize */
+			s.mb.RecordGpuDcgmCodecEncoderUtilizationDataPoint(now, encUtil)
+		}
+		if metric, ok := metrics["DCGM_FI_DEV_DEC_UTIL"]; ok {
+			decUtil := float64(metric.asInt64()) / 100.0 /* normalize */
+			s.mb.RecordGpuDcgmCodecDecoderUtilizationDataPoint(now, decUtil)
+		}
+		if metric, ok := metrics["DCGM_FI_DEV_FB_FREE"]; ok {
+			bytesFree := 1e6 * metric.asInt64() /* MBy to By */
+			s.mb.RecordGpuDcgmMemoryBytesUsedDataPoint(now, bytesFree, metadata.AttributeGpuMemoryStateFree)
+		}
+		if metric, ok := metrics["DCGM_FI_DEV_FB_USED"]; ok {
+			bytesUsed := 1e6 * metric.asInt64() /* MBy to By */
+			s.mb.RecordGpuDcgmMemoryBytesUsedDataPoint(now, bytesUsed, metadata.AttributeGpuMemoryStateUsed)
+		}
+		if metric, ok := metrics["DCGM_FI_DEV_FB_RESERVED"]; ok {
+			bytesFree := 1e6 * metric.asInt64() /* MBy to By */
+			s.mb.RecordGpuDcgmMemoryBytesUsedDataPoint(now, bytesFree, metadata.AttributeGpuMemoryStateReserved)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_DRAM_ACTIVE"]; ok {
+			s.mb.RecordGpuDcgmMemoryBandwidthUtilizationDataPoint(now, metric.asFloat64())
+		} else if metric, ok := metrics["DCGM_FI_DEV_MEM_COPY_UTIL"]; ok { // fallback
+			memCopyUtil := float64(metric.asInt64()) / 100.0 /* normalize */
+			s.mb.RecordGpuDcgmMemoryBandwidthUtilizationDataPoint(now, memCopyUtil)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_PCIE_TX_BYTES"]; ok {
+			pcieTx := int64(float64(metric.asInt64()) * (s.config.CollectionInterval.Seconds())) /* rate to delta */
+			s.mb.RecordGpuDcgmPcieIoDataPoint(now, pcieTx, metadata.AttributeNetworkIoDirectionTransmit)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_PCIE_RX_BYTES"]; ok {
+			pcieRx := int64(float64(metric.asInt64()) * (s.config.CollectionInterval.Seconds())) /* rate to delta */
+			s.mb.RecordGpuDcgmPcieIoDataPoint(now, pcieRx, metadata.AttributeNetworkIoDirectionReceive)
+		}
+		if metric, ok := metrics["DCGM_FI_PROF_NVLINK_TX_BYTES"]; ok {
+			nvlinkTx := int64(float64(metric.asInt64()) * (s.config.CollectionInterval.Seconds())) /* rate to delta */
```

Comment:
This only works if `scrape()` is called exactly once per collection interval. 

It sounds like the API already gives you timestamps for each point, so you can walk the list of points and integrate each point with the corresponding timestamp. And you probably need to do that from a separate collection goroutine if it needs to happen more often than the collection interval.

Reply:
Right now, `scrape()` _is_ called exactly once. I agree that the accuracy of the current approach is somewhat suboptimal, but I did fix a few bugs, and wasn't able to get the separate collection goroutine working. We can revisit this on the [igorpeshansky-dcgm-new-metrics-parallel](https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/compare/igorpeshansky-dcgm-new-metrics...igorpeshansky-dcgm-new-metrics-parallel) branch, but for now I'll stay with one scrape per collection.